templates/seL4VMParameters.template.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023, DornerWorks
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <camkes.h>
+#include <arm_vm/gen_config.h>
+
+/*- set vm_address_config = configuration[me.name].get('vm_address_config') -*/
+/*- set vm_image_config = configuration[me.name].get('vm_image_config') -*/
+/*- set linux_address_config = configuration[me.name].get('linux_address_config') -*/
+/*- set linux_image_config = configuration[me.name].get('linux_image_config') -*/
+
+/*- if vm_address_config -*/
+
+unsigned long ram_base = /*? vm_address_config.get('ram_base') ?*/;
+unsigned long ram_paddr_base = /*? vm_address_config.get('ram_paddr_base') ?*/;
+unsigned long ram_size = /*? vm_address_config.get('ram_size') ?*/;
+unsigned long ram_offset = /*? vm_address_config.get('ram_offset') ?*/;
+unsigned long dtb_addr = /*? vm_address_config.get('dtb_addr') ?*/;
+unsigned long initrd_max_size = /*? vm_address_config.get('initrd_max_size') ?*/;
+unsigned long initrd_addr = /*? vm_address_config.get('initrd_addr') ?*/;
+
+/*- if vm_address_config.get('kernel_entry_addr') != '-1' -*/
```

Comment:
It needs to be set or else it will cause an KeyError during the `init-build.sh` script. So keeping the -1 default value allows for the automatic calculation to function if its not set by the user.

Reply:
No sure I follow, wan't the point that we want the user to set this explicitly in the configuration, so it's more transparent what happens and there is less magic at the bottom? We could have an explicit linux template that provides defaults for Linux VMs, so users can use this one. Or a flag that says it's a linux kernel, which then enabled these well known defaults and maybe more convenince features. Also, it could be addressed in a meta VM creation system, that has more insight about the actual VMs, like the "vm composer" you have.

Reply:
I'm saying that every member of the vm_address_config struct _must_ have a value or the init-build.sh script will fail. We want to encourage the users to explicitly set the value, so when it is not set, CAmkES calculates a default, based off linux's offset, and the following warning will appear:

```
[258/305] Building C object CMakeFiles/vm0.instance.bin.dir/vm0/seL4VMParameters.template.c.obj
vm0/seL4VMParameters.template.c:27:2: warning: #warning Entry address has been calculated. Please use vm_address_config.kernel_entry_addr [-Wcpp]
   27 | #warning Entry address has been calculated. Please use vm_address_config.kernel_entry_addr
      |  ^~~~~~~
```

We need the `-1` default value because if the user didn't set the value, the configuration wouldn't even be able to compile. The `#warning` is what encourages the user to set the value.

Reply:
Ah, I see. So I wonder, should the complete macro `VM_INIT_DEF()` be deprecated then? It is creating a `vm_address_config` and a `linux_address_config` block now, which seems a bit odd. Is it somehow possible to not have it create a `vm_address_config` block and then print a warning if that it's outdated? And there is another new macro (maybe `VM_GENERIC_INIT_DEF()`) to be used that creates only a `vm_address_config` now. And if this new macro is used, the all fields must be set by the user?

Reply:
I think if we have a version, like Hannu was suggesting, we can only include the necessary attribute. So for now this is okay, but we can make that part of the next discussion.